internal/nostr/nostr.go

```diff
@@ -321,78 +320,76 @@ func (svc *Service) NIP47Handler(c echo.Context) error {
+			NostrId: requestData.SignedEvent.ID,
+			Content: requestData.SignedEvent.Content,
+		}
+	
+		err := svc.db.Create(&requestEvent).Error
+		if err != nil {
+			return c.JSON(http.StatusInternalServerError, ErrorResponse{
+				Message: "Failed to store request event",
+				Error:   err.Error(),
+			})
+		}
 	}
 
-	svc.Logger.WithFields(logrus.Fields{
-		"eventId":      requestData.SignedEvent.ID,
-		"walletPubkey": requestData.WalletPubkey,
-		"relayUrl":     requestData.RelayUrl,
-		"webhookUrl":   requestData.WebhookUrl,
-	}).Info("Storing request event and subscription")
-
-	subscription = Subscription{
-		RelayUrl:   requestData.RelayUrl,
-		WebhookUrl: requestData.WebhookUrl,
-		Open:       true,
-		Authors:    &[]string{requestData.WalletPubkey},
-		Kinds:      &[]int{NIP_47_RESPONSE_KIND},
-		Tags:       &nostr.TagMap{"e": []string{requestData.SignedEvent.ID}},
-		Since:      time.Now(),
-		Limit:      1,
-	}
-
-	err := svc.db.Create(&subscription).Error
-	if err != nil {
-		return c.JSON(http.StatusInternalServerError, ErrorResponse{
-			Message: "Failed to store subscription",
-			Error:   err.Error(),
-		})
-	}
-
-	requestEvent = RequestEvent{
-		NostrId: requestData.SignedEvent.ID,
-		Content: requestData.SignedEvent.Content,
-		SubscriptionId: subscription.ID,
-	}
-
-	err = svc.db.Create(&requestEvent).Error
-	if err != nil {
-		return c.JSON(http.StatusInternalServerError, ErrorResponse{
-			Message: "Failed to store request event",
-			Error:   err.Error(),
-		})
+	subscription := Subscription{
+		RelayUrl:     requestData.RelayUrl,
+		WebhookUrl:   requestData.WebhookUrl,
+		Open:         true,
+		Authors:      &[]string{requestData.WalletPubkey},
+		Kinds:        &[]int{NIP_47_RESPONSE_KIND},
+		Tags:         &nostr.TagMap{"e": []string{requestData.SignedEvent.ID}},
+		Since:        time.Now(),
+		Limit:        1,
+		RequestEvent: requestData.SignedEvent,
+		RequestID:    requestEvent.ID,
+		EventChan:    make(chan *nostr.Event, 1),
 	}
 
 	if subscription.WebhookUrl != "" {
-		go func() {
-			ctx, cancel := context.WithTimeout(svc.Ctx, 90*time.Second)
-			defer cancel()
-			event, _, err := svc.processRequest(ctx, &subscription, &requestEvent, &requestData)
-			if err != nil {
-				svc.Logger.WithError(err).Error("Failed to process request for webhook")
-				// what to pass to the webhook?
-				return
-			}
-			svc.postEventToWebhook(event, requestData.WebhookUrl)
-		}()
+		go svc.startSubscription(svc.Ctx, &subscription)
 		return c.JSON(http.StatusOK, NIP47Response{
 			State: "WEBHOOK_RECEIVED",
 		})
 	}
 
 	ctx, cancel := context.WithTimeout(c.Request().Context(), 90*time.Second)
 	defer cancel()
-	event, httpStatusCode, err := svc.processRequest(ctx, &subscription, &requestEvent, &requestData)
-	if err != nil {
-		return c.JSON(httpStatusCode, ErrorResponse{
-			Message: "Failed to process nip-47 request",
-			Error:   err.Error(),
+	go svc.startSubscription(ctx, &subscription)
+
+	select {
+	case <-ctx.Done():
```

Comment:
we could have a case for timeout here, then you can know exactly if it's a timeout or the request was cancelled

Reply:
Done!